service/worker/scanner/executions/shard_validator.go

```diff
@@ -0,0 +1,358 @@
+		executionValidationResult = executionValidator.validate(executionData)
+		if !executionValidationResult.isValid {
+			if result == nil {
+				result = newExecutionValidationFailure(executionData)
+			}
+			if result.Failures == nil {
+				result.Failures = map[string]executionValidatorResult{}
+			}
+			result.Failures[executionValidator.validatorTag()] = executionValidationResult
+		}
+	}
+	return result
+}
+
+// Can return NoFailure, nil, nil if execution was removed together with history branch.
+func fetchHistoryBranch(
+	executionInfo *persistencespb.WorkflowExecutionInfo,
+	executionState *persistencespb.WorkflowExecutionState,
+	shardID int32,
+	limiter quotas.RateLimiter,
+	historyDB persistence.HistoryManager,
+	totalDBRequests *int64,
+	execMgr persistence.ExecutionManager,
+) (executionValidatorResult, []*historypb.HistoryEvent, *persistencespb.HistoryBranch) {
+	var branch *persistencespb.HistoryBranch
+	currentVersionHistory, err := versionhistory.GetCurrentVersionHistory(executionInfo.VersionHistories)
+	if err != nil {
+		return executionValidatorResult{
+			isValid:               false,
+			failureReasonTag:      "failed_fetch_history",
+			failureReasonDetailed: "Failed to fetch current version history",
+		}, nil, nil
+	}
+
+	branch, err = serialization.HistoryBranchFromBlob(currentVersionHistory.BranchToken,
+		enumspb.ENCODING_TYPE_PROTO3.String())
+
+	if err != nil {
+		return executionValidatorResult{
+			isValid:               false,
+			failureReasonTag:      "failed_decode_branch_token",
+			failureReasonDetailed: "Failed to decode current branch token.",
+		}, nil, nil
+	}
+
+	var historyPageSize int = 1
+	readHistoryBranchReq := &persistence.ReadHistoryBranchRequest{
+		MinEventID:    common.FirstEventID,
+		MaxEventID:    common.EndEventID,
+		BranchToken:   currentVersionHistory.BranchToken,
+		ShardID:       &shardID,
+		PageSize:      historyPageSize,
+		NextPageToken: nil,
+	}
+
+	preconditionForDBCall(totalDBRequests, limiter)
+	history, err := historyDB.ReadHistoryBranch(readHistoryBranchReq)
+
+	if err != nil {
+		if errors.Is(err, gocql.ErrNotFound) {
+			stillExists, failureInfo := concreteExecutionExists(executionInfo, executionState, execMgr, limiter, totalDBRequests)
+			if !failureInfo.isValid {
+				return failureInfo, nil, nil
+			}
+			if !stillExists {
+				return executionValidatorResultNoCorruption, nil, nil
+			}
+			return executionValidatorResult{
+				isValid:               false,
+				failureReasonTag:      "invalid_missing_corresponding_branch",
+				failureReasonDetailed: "Execution misses corresponding branch entry.",
+				error:                 nil,
+			}, nil, nil
+		}
+		return executionValidatorResult{
+			isValid:               false,
+			failureReasonTag:      "invalid_missing_corresponding_branch",
+			failureReasonDetailed: "Failed to fetch history branch with Error other than gocql.ErrNotFound.",
+			error:                 err,
+		}, nil, nil
```

Comment:
above checks `gocql.ErrNotFound`, meaning this error result will be returned if seeing NOT gocql.ErrNotFound, which also includes timeout / internal server errors.